### PR TITLE
Manipulator with extra detail modes, display features and sort options.

### PR DESCRIPTION
**Pull request by strainer**
I dont know if this is acceptable because its a barrel load of features in one, but Im proud of this work and hope it can be accommodated.

It includes 2 new detail views, to display unit attributes and to report unit 'Notices' which are things like needs doctor, needs recovery, has syndrome etc.
It also includes a color themed hinting system which reveals which labors each unit will be good/fast at. The color hinting is off by default and toggleable. Its designed to be the default mode easy on the eye, but the b&w mode may be necessary for some custom color pallets.

It includes new sort options - by skill groups of kind "martial", "scholar","perform" and "civil" (all the normal labors like masonry etc). An "in demand" sort option which gives a rough idea of which units have many/few roles enabled.

The sort settings distinguish between 'group sorts', like squads, selection and 'arrival group' and 'fine sorts' like the skill class sorts which do not form groups. The group sort is selected independanty in combination with the fine sort, and the groups are indicated visually by alternating color of names in the name listing. Ive tried to make the new sort controls obvious and easy to toggle through.

The mouse controls have been updated for the new features and other interface tweaks.

The code will want put through a formatter for the repos style guide, and maybe restructuring since I have little experience of cpp idioms. Ive tried to keep it tidy and should find time to format and improve it myself in the future, but Im pushing now on the chance other curious devs might help, if it needs it.
Its quite well play tested on df43 linux, and it seems stable so, I hope its worth a look : )

### CODE REVIEW

**Comment by strainer** [comment]
Apologies - I will fix some comments and things these evening.

**Comment by strainer** [comment]
The fundamentals of manipulator are great, plenty of good work invested into setting up the basic layout and the nicknaming facility is invaluable. It had some very noticable issues. The stress display was colored as though postive values are a bad thing, and the game engines raw integer value was listed without scaling it, making a listing of eg. -50,000 stress seem huge, while it is actually not very large. The displayed stress was even capped at -99,999 which in the game engine is early days as far as dwarf stress is concerned. I have noticed dwarfs start developing syndromes once they get below around -200,000  stress - thats twice as low as manipulator would display. 

I cleared this up by just displaying the integer stress counter divided by 100. I kept the friendly green coloring of bad stress levels for comfort sake, but made them turn brass, steel and adamantine for those attentive enough to make dwarves happy (ive only got to brass myself)

The sorting system also seemed not quite finished, not being able to subsort arrivals and difficult to select, orientate and then apply the right sorts, leave the screen by accident and the settings were lost.

Ive put over 50 hours into polishing all these issues away. The only thing I can see left that would be missed is a dwarf name search/jump feature, which wont be terribly difficult to add sometime...

**Comment by strainer** [comment]
To correct my previous message - the stress score was correctly colored = naturally. I will revert that change. Seems to hardly matter since I tortured a fort for years and nairly a unit is positively stressed. Thinking about translating it to a different meter like "myrth" or "grit"

**Comment by Clyybber** [comment]
Did you test it on 44.0x yet?

File: plugins/manipulator.cpp [hunk 1]
```diff
@@ -2164,14 +3807,12 @@ void viewscreen_unitlaborsst::render()
     OutputString(10, x, y, Screen::getKeyDisplay(interface_key::CUSTOM_P));
     OutputString(15, x, y, ": Apply Profession ");
     OutputString(10, x, y, Screen::getKeyDisplay(interface_key::CUSTOM_SHIFT_P));
-    OutputString(15, x, y, ": Save Prof. ");
+    OutputString(15, x, y, ": Read Prof. ");
 }
 
 df::unit *viewscreen_unitlaborsst::getSelectedUnit()
 {
-    // This query might be from the rename plugin
```
Comment (lethosor) [inline plugins/manipulator.cpp]: This is relevant - it's saying that the rename plugin could have changed some unit's name, so names should be refreshed.

File: plugins/manipulator.cpp [hunk 2]
```diff
@@ -2164,14 +3807,12 @@ void viewscreen_unitlaborsst::render()
     OutputString(10, x, y, Screen::getKeyDisplay(interface_key::CUSTOM_P));
     OutputString(15, x, y, ": Apply Profession ");
     OutputString(10, x, y, Screen::getKeyDisplay(interface_key::CUSTOM_SHIFT_P));
-    OutputString(15, x, y, ": Save Prof. ");
+    OutputString(15, x, y, ": Read Prof. ");
```
Comment (lethosor) [inline plugins/manipulator.cpp]: I haven't looked around a lot, but I don't see anything dealing with the CUSTOM_SHIFT_P key that you changed, so I don't think changing this text is accurate. Let me know if I'm wrong.

File: plugins/manipulator.cpp (hunk 1 above)
Comment (lethosor) [inline plugins/manipulator.cpp]: This is relevant - it's saying that the rename plugin could have renamed this unit, so names should be refreshed.

File: plugins/manipulator.cpp (hunk 2 above)
Comment (lethosor) [inline plugins/manipulator.cpp]: I haven't looked around a lot, but I don't see anything dealing with the CUSTOM_SHIFT_P key that you changed, so I don't think changing this text is accurate. Let me know if I'm wrong.

**Comment by lethosor** [comment]
I haven't really gotten around to looking at this much. I suppose I should at some point. It's a large enough diff that it's hard to check on mobile, so I haven't gone through all of it.

That said, there are some unrelated changes in here, and a lot of style issues. Granted, there are a decent number of plugins with style issues already, but manipulator was mostly okay before, so I'd rather not introduce these. I'll go through and leave some comments.

File: plugins/manipulator.cpp [hunk 3]
```diff
@@ -285,14 +297,14 @@ struct UnitInfo
     bool allowEdit;
     string name;
     string transname;
+    string lastname;
```
Comment (lethosor) [inline plugins/manipulator.cpp]: Can't you just get this from unit->name?

File: plugins/manipulator.cpp [hunk 4]
```diff
@@ -304,144 +316,372 @@ struct UnitInfo
...
 };
-enum altsort_mode {
-    ALTSORT_NAME,
-    ALTSORT_SELECTED,
-    ALTSORT_DETAIL,
-    ALTSORT_STRESS,
-    ALTSORT_ARRIVAL,
-    ALTSORT_MAX
+const char * const detailmode_shortnames[] = {
+  "Profession,  ",
```
Comment (lethosor) [inline plugins/manipulator.cpp]: I guess this works, but it'd be better to just store the strings you need and add spaces/punctuation where you need them. (Maybe our stl_sprintf() would help, or some variation on OutputString(), not sure.)

File: plugins/manipulator.cpp [hunk 5]
```diff
@@ -304,144 +316,372 @@ struct UnitInfo
...
+  "Squads,      ",
+  "Actions,     ",
+  "Notices,     ",
+  "Attributes,  ",
+  "Oops"
```
Comment (lethosor) [inline plugins/manipulator.cpp]: Ideally the "Oops" strings wouldn't be necessary

File: plugins/manipulator.cpp [hunk 6]
```diff
@@ -304,144 +316,372 @@ struct UnitInfo
...
+  "Martial",
+  "Perform",
+  "Scholar",
+  "Stress",
+  "Notices",
+  "Oops",
+};
+
+static string cur_world;
+static int detail_mode=0;
```
Comment (lethosor) [inline plugins/manipulator.cpp]: There's some very inconsistent spacing going on in a lot of places. This section in particular has between 0 and 2 spaces around `=` signs.

File: plugins/manipulator.cpp [hunk 7]
```diff
@@ -1253,150 +2309,396 @@ void viewscreen_unitlaborsst::refreshNames()
...
+
+    if(first_column>=NUM_LABORS-row_width)
+        first_column=NUM_LABORS-row_width;
 
-    // if it shrinks horizontally, scroll to the right to keep the cursor visible
-    if (first_column < sel_column - col_widths[DISP_COLUMN_LABORS] + 1)
-        first_column = sel_column - col_widths[DISP_COLUMN_LABORS] + 1;
+}
+
+bool viewscreen_unitlaborsst::scrollknock(int *reg, int stickval, int passval){
```
Comment (lethosor) [inline plugins/manipulator.cpp]: Is this supposed to wrap around after a certain time or something?

File: plugins/manipulator.cpp [hunk 8]
```diff
@@ -542,94 +782,782 @@ class StringFormatter {
...
-    string get_last_name(UnitInfo *u)
-    {
-        df::language_name name = u->unit->name;
-        string ret = "";
-        for (int i = 0; i < 2; i++)
+static bool labor_skill_map_made = false;
+
+namespace unit_info_ops{
+
+    //these are copied from managelabor.cpp ( skill_attr_weight )
```
Comment (lethosor) [inline plugins/manipulator.cpp]: labormanager? Might be better to move to core if multiple things need it, but not too necessary.

File: plugins/manipulator.cpp [hunk 9]
```diff
@@ -304,144 +316,372 @@ struct UnitInfo
...
-        }
-        else
-        {
-            if (l1 != l2)
-                return l1 < l2;
-            if (e1 != e2)
-                return e1 < e2;
-        }
+        l1 = unitSkillRating(d1,sort_skill)*50;
+        l1 += 25+unitSkillExperience(d1,sort_skill)*40/(skill_levels[l1/50].points+1);
```
Comment (lethosor) [inline plugins/manipulator.cpp]: No spaces

File: plugins/manipulator.cpp [hunk 10]
```diff
@@ -542,94 +782,782 @@ class StringFormatter {
...
+        if( (unit->curse.add_tags1.bits.OPPOSED_TO_LIFE || unit->curse.add_tags1.bits.NOT_LIVING)
+            && !unit->curse.add_tags1.bits.BLOODSUCKER )
+            cursetype = "Zmbie ";
+
+        if(!unit->curse.add_tags1.bits.NOT_LIVING
+            && unit->curse.add_tags1.bits.NO_EAT
+            && unit->curse.add_tags1.bits.NO_DRINK
+            && unit->curse.add_tags2.bits.NO_AGING
+            )
+            cursetype = "Ncrmncr ";
```
Comment (lethosor) [inline plugins/manipulator.cpp]: Do these need to be abbreviated?

File: plugins/manipulator.cpp [hunk 11]
```diff
@@ -1108,45 +2039,91 @@ class viewscreen_unitlaborsst : public dfhack_viewscreen {
...
+    }
+
+    return unit;
+}
+
+int viewscreen_unitlaborsst::findUnitsListPos(int unit_row){
+
+    int ULcurpos = -1;
+    if (VIRTUAL_CAST_VAR(unitlist, df::viewscreen_unitlistst, parent))
+    for (int i = 0; i < unitlist->units[unitlist->page].size(); i++)
```
Comment (lethosor) [inline plugins/manipulator.cpp]: Inconsistent indentation

File: plugins/manipulator.cpp [hunk 12]
```diff
@@ -1173,43 +2150,118 @@ viewscreen_unitlaborsst::viewscreen_unitlaborsst(vector<df::unit*> &src, int cur
...
+        first_row = units.size() - display_rows;
 
     last_selection = -1;
+
+    read_manipulator_config();
+}
+
+void viewscreen_unitlaborsst::calcArrivals()
+{
+    int bi=-100;
```
Comment (lethosor) [inline plugins/manipulator.cpp]: What do these names mean?

File: plugins/manipulator.cpp [hunk 13]
```diff
@@ -1253,150 +2309,396 @@ void viewscreen_unitlaborsst::refreshNames()
...
+
+            int dm=0;
+            if(cur->unit->status.labors[cur_labor]){
+                if(cur_skill==df::job_skill::NONE){
+                    if(cur_labor==unit_labor::HAUL_STONE
+                     ||cur_labor==unit_labor::HAUL_WOOD
+                     ||cur_labor==unit_labor::HAUL_ITEM
+                     ||cur_labor==unit_labor::HAUL_STONE
+                     ||cur_labor==unit_labor::BUILD_CONSTRUCTION){
+                        dm=12;
```
Comment (lethosor) [inline plugins/manipulator.cpp]: ?

File: plugins/manipulator.cpp [hunk 14]
```diff
@@ -807,7 +1735,9 @@ class viewscreen_unitbatchopst : public dfhack_viewscreen {
         formatter.add_option("i", "Position in list", unit_ops::get_list_id);
         formatter.add_option("pi", "Position in list, among dwarves with same profession", unit_ops::get_list_id_prof);
         formatter.add_option("gi", "Position in list, among dwarves in same profession group", unit_ops::get_list_id_group);
-        formatter.add_option("ri", "Raw unit ID", unit_ops::get_unit_id);
```
Comment (lethosor) [inline plugins/manipulator.cpp]: Why was this one removed? I think the other two are ones you added for debugging purposes, or something like that.

File: plugins/manipulator.cpp [hunk 15]
```diff
@@ -867,7 +1797,8 @@ class viewscreen_unitbatchopst : public dfhack_viewscreen {
             else if (events->count(interface_key::SELECT))
             {
                 apply((cur_page == NICKNAME) ? unit_ops::set_nickname : unit_ops::set_profname, entry, &formatter);
-                select_page(MENU);
+                Screen::dismiss(this);
```
Comment (lethosor) [inline plugins/manipulator.cpp]: This is somewhat of a personal preference. The original behavior here was intended to allow multiple changes to be made without having to re-enter the menu every time. Of course, the menu doesn't have many options now, so maybe it's not useful behavior. Just seems unrelated.

**Comment by lethosor** [comment]
To clarify, I think this looks interesting, and I'm not opposed to it in general. However, it's a lot of changes, not all of which I've had time to look over. Merging it requires us to be able to maintain it going forward if needed, so style/readability is somewhat important. (The downside to running it through some styler, though, is that it might change lines you didn't touch yourself, which makes diffs harder to read.)

I'm fairly busy for the next few days, but I'll try to look over it again when I can and reply.

File: plugins/manipulator.cpp (hunk 3 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Maybe i looked and anorher issue, i mean - cant you get name from unit->name too? yet name is in uinfo here. I cached last name here because it is accessed by a sort comparator.

File: plugins/manipulator.cpp (hunk 4 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Having customs strings for displayed labels can ne more convienient to reconfigure than custom spacing logic.

File: plugins/manipulator.cpp (hunk 5 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Its not necessary, never displayed. Its a joke about the corresponding _MAX enumeration which i found in the existing code and changer to _OVER.

File: plugins/manipulator.cpp (hunk 6 above)
Comment (strainer) [reply plugins/manipulator.cpp]: I probably need a linter. I have very poor eyesight and dont personally concern about whitespace.

File: plugins/manipulator.cpp (hunk 10 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Quite, if we prefer a few of them to be able to fit into the available detail space.

File: plugins/manipulator.cpp (hunk 14 above)
Comment (strainer) [reply plugins/manipulator.cpp]: I mistook it for one of my debug fields.

File: plugins/manipulator.cpp (hunk 15 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Few users will want to change both nickname and professions consequtively, and those who might do wont be repeating that operation often so the escape keypress involved matters little for them. But if you to go through the list nicking units one by one which is a natural action, its much nicer to not have an escape keypress in the way.

File: plugins/manipulator.cpp (hunk 12 above)
Comment (strainer) [reply plugins/manipulator.cpp]: I could comment them and/or call them single letters, but they are simply defined by their use in the code which very shortly follows their declaration here. They are bai :before active index, cai: current active index, bi before index, ci: current index

File: plugins/manipulator.cpp (hunk 13 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Maybe this is from difficulty reviewing on your phone. an 'in demand' score is being calculated here.

File: plugins/manipulator.cpp (hunk 7 above)
Comment (strainer) [reply plugins/manipulator.cpp]: It requires a double tap to wrap around. This creates ergonomically comfortable edges, accidental and unexpected jumping is disorientating in user experience. Measuring a system time was necessary to implement this behaviour reliably.

File: plugins/manipulator.cpp (hunk 2 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Good catch, I did not intend to change that.

File: plugins/manipulator.cpp (hunk 1 above)
Comment (strainer) [reply plugins/manipulator.cpp]: Also, I mistook that for one of my own tips that I culled.

**Comment by strainer** [comment]
"I think this looks interesting"

Find 10 minutes to compile and look at it - it works nice.

Clyybber - Its not tried on the alpha or on windows yet, if you can compile you could maybe try it too - give it a look over?

Just one file to drop in - [raw link](https://raw.githubusercontent.com/strainer/dfhack/manipu_remix/plugins/manipulator.cpp)

**Comment by Clyybber** [comment]
Oh yeah, will do. Just need to works out how to compile it in Windows, since I mainly use linux

**Comment by strainer** [comment]
Thanks Clyybber - setting up new build system can sure be tricky though. I should ask on irc or something for someone with a windows setup who could make one in five minutes,,,

Lethosor thanks for the code review that was my first. I think ive got most of the formatting corrected except spaces in long formulas - if you are set on that I will put them in, I just feel those formulas are a bit more presentable without. I can see the benefit of correcting the assignment spacing and things.

**Comment by strainer** [comment]
Hooray! Thankyou for bearing with me Lethosor!  God I hope it doesnt crash 44 ... I will be paying attention for issues and feedbacks.

**Comment by strainer** [comment]
Modders can put a little theme file in manipulators config directory, and a couple of pertanent tweaks there. Tested on 44 03 a1

**Comment by strainer** [comment]
Cheatmode enabled by cycling backward through themes allows redistribution of jobskills, points need taken away before they can be added. A further cheatsetting allows unlimited skill adjustment.

**Comment by lethosor** [comment]
Are you changing the name of this or something?

File: plugins/manipulator.cpp [hunk 16]
```diff
@@ -304,144 +352,393 @@ struct UnitInfo
...
+    config_dfkeeper.ival(0) = color_mode+1;
+    config_dfkeeper.ival(1) = hint_power;
+    config_dfkeeper.ival(2) = spare_skill;
+    config_dfkeeper.ival(3) = show_details+1;
+    config_dfkeeper.ival(4) = tran_names;
+    config_dfkeeper.ival(5) = 0;
+    config_dfkeeper.ival(6) = 0;
+}
+
+void read_dfkeep_config()
```
Comment (Clyybber) [inline plugins/manipulator.cpp]: This doesn't really matter, but maybe rename read_dfkeep_config() to read_dfkeeper_config(), save_dfkeep_config() to save_dfkeeper_conifg() and dfkeep_palette to dfkeeper_palette?

**Comment by strainer** [comment]
As discussed in irc, changing name, need to do testing, wikiguide etc before pushing later revisions. The earlier revisions submitted would have been good as an upgrade to the legacy plugin, but since it has gotten this [bloated] or [featured] I dont mind it coming out of the milestone.

**Comment by Clyybber** [comment]
Why was this closed? Will it be made a standalone plugin?

**Comment by strainer** [comment]
Hi Clyybber, thanks for noticing :) - I closed it myself as the latest work as it stands seems to have no chance of being pulled without a lot of conversation and alteration for dfhacks preferences, even the earlier improvements which I made with open source maintainability in mind where dropped from the scheduled milestone release when I changed the name and warned that later revisions might need testing. So I decided to take ownership of my version and try to get it distributed and tested myself. dfhack can take it back and do whatever at any time but im not 'pushing' it uphill anymore. I think the latest version is ridiculously good if I say-so :) 

I put a compile that goes with the latest linux newbie pack here at bay12 - http://www.bay12forums.com/smf/index.php?topic=163211.msg7681706 ,there is a link to a description with screenshots there too.

I should get my hands on a windows machine eventually to make windows compiles, perhaps it wont break out of the 'looks interesting' box until then. If you ever get an hour or two to do a windows compile let me know, there is a recent change to df structures which requires two lines of my source to be set to fit ("anon_3" changes to "attitude").  Also if dfhacks compile instructions for windows work as well as the linux instructions, it could be very quickly done.

Thanks again for following Clyybber
